### PR TITLE
Update improve devcontainer

### LINKED ISSUE #37169: Update Dockerfile and docker-compose in .devcontainer directory
The current implementation of the Docker Image in/.devcontainer uses PHP7.4, which is no longer compatible with the minimal PHP requirement of PHP 8.0 for developers trying to set up a local development environment with docker. Updating (and maybe tweaking a bit) the current implementation would help developers interested in contributing to set up their local environment quicker. What do you think?

### PR DESCRIPTION
**Pull request by R0Wi**
* Resolves: #37169 

## Summary

Based on https://github.com/nextcloud/server/pull/37176.
Added many minor improvements to the DevContainer configuration so that one is now able to open the project in VSCode DevContainers / Codespaces and start working right away.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

## How to test

See `.devcontainer/README.md`. Tested on GitHub Codespaces, Docker for Linux and Docker Desktop for Windows. Just open, wait and you're done :smile:

### CODE REVIEW

**Review by szaimen — APPROVED** [review]
LGTM but didnt test

**Comment by R0Wi** [comment]
If anyone wants to check this out on Codespaces, it's easy, just open Codespaces directly on this branch

image

**Comment by DaphneMuller** [comment]
Wow, very nice work @R0Wi ! I just tried it out and completed the first tutorial from nextcloud.com/developer with your environment. This is amazing. It all runs without any pre-knowledge with just a few clicks.

**Review by artonge — APPROVED** [review]
@blizzz should know about excluding files from the release.

**Comment by blizzz** [comment]
We never used to commit editor configs though. Not convinced that Microsoft should get extra exposure.

**Comment by R0Wi** [comment]
In general I'd agree but starting a new devcontainer having a debug profile ready which matches the apache/cli config is quite convenient. Alternatively we could move it to the .devcontainer folder and copy it in the startup routine. But I'm not sure if VSCode needs to be reloaded then...

**Comment by octopusx** [comment]
Hello, I hope this comment is not too spammy, I was a part of the JB hackathon at the NC headquarters last week and was working on improving the docker-compose dev environment, and this is basically what I was looking for. Just tested this branch on my system and it works beautifully. As for the issue of committing an editor-specific config file into the main repository, I agree that it feels a little like playing favorites, but at the same time this setup is still docker-compose based and anyone should be able to run it with or without VSCode or Codespaces.

tldr: big thumbs up from me, hope this gets merged soon and I am looking forward to contributing to this repo soon :+1:

**Comment by octopusx** [comment]
Doing some further testing, I noticed that this command is used to bootstrap the NC server inside the container: https://github.com/nextcloud/server/pull/37521/files#diff-aaf9e7764a12876cbe70d492c719e21e8590390a38d0f037193a765df858fdbfR18

Once the bootstrapping is done and the container is running though I am not able to use the `php occ` command, instead getting the following error:
```
devcontainer@8ea2a6470692:/var/www/html$ php ./occ help
Xdebug: [Step Debug] Could not connect to debugging client. Tried: localhost:9003 (through xdebug.client_host/xdebug.client_port) :-(
An unhandled exception has been thrown:
OCP\HintException: [0]: Memcache \OC\Memcache\APCu not available for local cache (Is the matching PHP module installed and enabled?)
```
 So to run it I need to specify the `apc.enable_cli=1` argument, like so:
```
sudo -u ${APACHE_RUN_USER} php --define apc.enable_cli=1 occ
```
In the current setup everything in the `/etc/php/8.1/*` path is read-only, so I can not add this permanently as a config line.

Does anyone else have this issue, and if so, could we make this a part of the bootstrapping for this dev environment? Happy to help troubleshoot.

**Comment by R0Wi** [comment]
Thanks for your feedback, @octopusx . Just added the `apcu`-line to the `Dockerfile`, this should fix your problem.

@blizzz please let me know your preferred way to deal with the `launchsettings.json`. If you like I'll move them to the `.devcontainer` and copy it to the correct location on container startup.

**Comment by blizzz** [comment]
> Thanks for your feedback, @octopusx . Just added the `apcu`-line to the `Dockerfile`, this should fix your problem.
> 
> @blizzz please let me know your preferred way to deal with the `launchsettings.json`. If you like I'll move them to the `.devcontainer` and copy it to the correct location on container startup.

Moving it would be great, thank you!

**Comment by R0Wi** [comment]
@blizzz @artonge @szaimen I think all issues should be resolved now. Failing PR tests seem not to be related to the code changes, any chance to get this merged even though some tests are failing?

**Comment by szaimen** [comment]
CI failure indeed seems unrelated
